alexandria/core/tasks.py

```diff
@@ -0,0 +1,55 @@
+import hashlib
+from pathlib import Path
+
+import tika.language
+import tika.parser
+from django.conf import settings
+from django.contrib.postgres.search import SearchVector
+from django.db.models import Value
+
+from alexandria.core.models import File
+from celery import shared_task
+
+
+@shared_task(soft_time_limit=301)
+def set_content_vector(file_pk: str):
+    file = File.objects.get(pk=file_pk)
+    file.content.file.file.seek(0)
+
+    # tika has an internal time limit of 300s, set the request limit to match that
+    # different values should be set in tika as well
+    # https://github.com/CogStack/tika-service/blob/master/README.md#tika-parsers-configuration
+    parsed_content = tika.parser.from_buffer(
+        file.content.file.file, requestOptions={"timeout": 300}
+    )
+
+    name_vector = SearchVector(Value(Path(file.name).stem), weight="A")
+    if not parsed_content["content"]:
+        file.content_vector = name_vector
+        file.save()
+        return
+
+    # use part of content for language detection, beacause metadata is not reliable
+    file.language = tika.language.from_buffer(parsed_content["content"][:1000])
+    config = settings.ALEXANDRIA_ISO_639_TO_PSQL_SEARCH_CONFIG.get(
+        file.language, "simple"
+    )
+    file.content_vector = name_vector + SearchVector(
+        Value(parsed_content["content"].strip()),
+        config=config,
+        weight="B",
+    )
+
+    file.save()
```

Comment:
Something just popped into my mind: We're saving the whole file object here, potentially overwriting the checksum that may now run in parallel (and probably finished earlier).

Either we'll need to wrap the task into `@transaction.atomic` or explicitly update only the `content_vector` field. I'd prefer the latter:

```suggestion
    # Update only content vector, to avoid race with the checksum task
    File.objects.filter(pk=file.pk).update(content_vector=file.content_vector)
```

Reply:
Makes sense, didnt consider the overwriting each other. Applied